### PR TITLE
Add a batch-waitall in MPI_Waitall to reduce cache misses

### PR DESCRIPTION
**Pull request by danghvu**
This is based on PR #2742.
It improves the performance when dealing with large number of requests (by reducing cache misses with the cost of more instructions). It also removes an memory allocation, because the batch size is set to be the available stack buffers for internal requests.

This is WIP, any feedback to improve is appreciated.
(I also fixed the merge conflicts, see only this 3c7411e commit to track changes vs. original PR)
```
window=8192
<2742>
size | MB/s | Messages/s | TOT_INS | L1_DCM
1 | 0.77 | 772205.74 | 1204033.79 | 5069.38
2 | 1.53 | 766035.08 | 2408041.65 | 10116
4 | 2.98 | 746067.09 | 3612049.53 | 15103.13
8 | 6.10 | 761932.77 | 4816057.53 | 20157.45

<this PR>
size | MB/s | Messages/s | TOT_INS | L1_DCM
1 | 0.80 | 798943.71 | 1230194.72 | 1691.25
2 | 1.59 | 792737.86 | 2460361.73 | 3411.09
4 | 3.15 | 786491.78 | 3690528.71 | 5141.10
8 | 6.34 | 792034.11 | 4920695.91 | 6860.15

<master>
size | MB/s | Messages/s | TOT_INS | L1_DCM
1 | 0.78 | 779323.83 | 1163436.04 | 2781.15
2 | 1.56 | 780292.07 | 2326846.47 | 5629.61
4 | 3.15 | 786695.38 | 3490256.54 | 8484.65
8 | 6.34 | 791960.40 | 4653666.75 | 11321.55

window=64 (minimal difference)
<master>

Size | MB/s | Messages/s | TOT_INS | L1_DCM
1 | 0.89 | 886386.69 | 9940.47 | 23.20
2 | 1.81 | 904826.96 | 19853.69 | 46.15
4 | 3.67 | 917355.87 | 29766.89 | 67.70
8 | 7.30 | 911918.77 | 39680.12 | 88.24

<this PR>
Size | MB/s | Messages/s | TOT_INS | L1_DCM
1 | 0.90 | 900083.26 | 10193.47 | 20.20
2 | 1.78 | 890365.15 | 20359.67 | 40.70
4 | 3.68 | 919155.49 | 30525.87 | 62.06
8 | 7.37 | 921270.72 | 40692.07 | 82.91
```

### CODE REVIEW

**Comment by pavanbalaji** [comment]
Thanks, @danghvu.  Can you make this PR rebase-clean on top of `mpich/master`?  It's hard to follow the patches along in this PR.

**Comment by pavanbalaji** [comment]
@halimamer can you work with @danghvu to clean up this PR?

@danghvu some of @hajimefu's patches seem to be mixing multiple things together into a single patch.  That's discouraged.  Can you split them out into independent pieces?  For example, this patch:

```
MPI: Change MPIR_Wait*_impl semantics

This patch changes MPIR_Wait*_impl (Wait/Waitall/Waitany) semantics in
several ways.
* Now take MPIR_Request * pointer instead of MPI_Request handle
* Do not check for MPI_REQUEST_NULL. It is caller's responsibility
  to ensure the request pointer is valid.
* Do not free the request. They simply make progress until
  the request completes. A caller should then call MPIR_Request_complete
  to free the object and clear the handle.
  (In some minor cases MPIR_Wait*_impl still sets the status,
   for example for generalized request.)
```

**Comment by danghvu** [comment]
@pavanbalaji I tried to rebase at first but it has a bunch of rebase conflicts which is really painful to fix. That's why I resort to a merge which I only have to fix a few places.

Can you give me an advice on what best to do ? Maybe reworking it with my own commit in a separate PR (moving his code over manually) is better ? Do I need to retain his commit reference / authorship ?

**Comment by pavanbalaji** [comment]
Explicitly picking code is fine.  If most of the code is copied from @hajimefu's PR, it's best to retain his ownership.  If most of the code is yours, you can ignore that.  Frankly, most of us don't really care about commit authorship, but I don't want to assume that about @hajimefu.  Authorship is mainly to know whom to blame when something goes wrong.  :-)

**Comment by danghvu** [comment]
Rebased + Reworked to split large commits. It's difficult to split further in those commits that change the `MPIR_*_impl` function to take `MPIR_Request` because other stuff depends on it.

**Comment by pavanbalaji** [comment]
@danghvu the branch name used for this PR is `pr2742`, and is very misleading.  You might want to fix that.  When we checkout your branch, we won't know which PR we are working on.

File: src/mpi/request/test.c [hunk 1]
```diff
@@ -154,11 +134,32 @@ int MPI_Test(MPI_Request *request, int *flag, MPI_Status *status)
...
+        *flag = TRUE;
+        goto fn_exit;
+    }
+
     /* ... body of routine ...  */
 
     mpi_errno = MPIR_Test_impl(request, flag, status);
     if (mpi_errno) goto fn_fail;
     
+    if (MPIR_Request_is_complete(request_ptr)) {
```
[the remark this answers is not in the document]
Comment (hajimefu) [reply src/mpi/request/test.c]: Yeah, we can check `flag` instead.

File: src/mpi/request/test.c [hunk 2]
```diff
@@ -154,11 +134,32 @@ int MPI_Test(MPI_Request *request, int *flag, MPI_Status *status)
...
+        mpi_errno = MPIR_Request_complete(request, request_ptr, status,
+                                          &active_flag);
```
[the remark this answers is not in the document]
Comment (hajimefu) [reply src/mpi/request/test.c]: `MPIR_Request_complete` frees the MPIR request object, sets `MPI_REQUEST_NULL` to the user request, and sets the status and the flag.

This may be something we want to consider renaming. The current naming of these functions are confusing. How about `MPIR_Request_post_complete` for something?

File: src/mpi/request/test.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (danghvu) [reply src/mpi/request/test.c]: You are right, I think @hajimefu just systematically moved the code out. We can check the flag instead of calling a function here.

File: src/mpi/request/test.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (danghvu) [reply src/mpi/request/test.c]: This is the same as the previous code, we have to check if it is really completed first then call `Request_complete` which really does the cleanup.

[the remark this answers is not in the document]
Comment (pavanbalaji) [reply src/mpi/request/test.c]: @hajimefu `post_complete` is an awful name.  It doesn't give any useful information.

Besides, I don't think we need to complete the request again.  In fact, I'd argue that you should *not* complete the request again (since the `cc` has already hit zero).  You should simply free it.

**Comment by hajimefu** [comment]
@danghvu Thank you for taking a crack on this. As for authorship, I think you can use something like "Co-authored-by: " in the commit message if needed. As @pavanbalaji mentioned, it would help finding who would be responsible for the change later (e.g. ask questions, etc.)

File: src/mpi/request/test.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (pavanbalaji) [reply src/mpi/request/test.c]: I'd recommend a separate patch for cleaning that up as a prereq patch, instead of merging that piece with this patch.

File: src/mpi/request/test.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (pavanbalaji) [reply src/mpi/request/test.c]: Alternatively, if `MPIR_Test_impl` is only needed for checking for generalized requests, we should just name it that, and remove the completion check for that request from there.

**Comment by danghvu** [comment]
@hajimefu Thank you, I will do surely do that in the next version of the rework (making a new PR) after gathering some feedbacks from @pavanbalaji

File: src/mpi/request/test.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (hajimefu) [reply src/mpi/request/test.c]: @pavanbalaji I have to admit I'm not good at naming...
Before going into that discussion, I want to make it clear that `MPIR_Request_complete` does *NOT* complete the request (=it does not decrement the completion counter). I think this is where all the confusion comes from, and this is why I'm proposing renaming it.

[the remark this answers is not in the document]
Comment (pavanbalaji) [reply src/mpi/request/test.c]: Oof.  `MPID_Request_complete` decrements the `cc`.  I agree that the name is just awful and should be renamed.

[the remark this answers is not in the document]
Comment (hajimefu) [reply src/mpi/request/test.c]: Okay. Now we are in the same page. Do you want to come up with a better name? `MPIR_Request_cleanup`?

[the remark this answers is not in the document]
Comment (pavanbalaji) [reply src/mpi/request/test.c]: In fact, the only thing it's not doing is "completing" the request.  :-)

[the remark this answers is not in the document]
Comment (danghvu) [reply src/mpi/request/test.c]: `MPIR_Request_finalize` ? If nothing better I will just leave it to move on first.

**Comment by wesbland** [comment]
Yeah, you just have to be careful about how you put it in the commit message. I'd recommend reading the blog post linked above to get it right.

File: src/mpi/request/testall.c [hunk 3]
```diff
@@ -35,148 +35,26 @@ int MPI_Testall(int count, MPI_Request array_of_requests[], int *flag,
 #define FUNCNAME MPIR_Testall_impl
 #undef FCNAME
 #define FCNAME MPL_QUOTE(FUNCNAME)
-int MPIR_Testall_impl(int count, MPI_Request array_of_requests[], int *flag,
+int MPIR_Testall_impl(int count, MPIR_Request *request_ptrs[], int *flag,
```
Comment (danghvu) [inline src/mpi/request/testall.c]: @pavanbalaji @hajimefu Do we need `flag` here? This code does not use it if I read correctly. Is it just to comply with other `MPIR_*_impl` series having the same header as `MPI_*` except `MPIR_Request` is used ?

Comment (hajimefu) [reply src/mpi/request/testall.c]: I think you're right. We can remove it. Apparently we blindly copied the existing function signature and did not carefully look at each parameter.

Comment (danghvu) [reply src/mpi/request/testall.c]: Actually, the better question is whether the associated device call is going to set the flag (is the "device" layer supposed to set the user flag ?) since this is going to be turn into `MPID_Testall`, if yes then we have to keep it.

**Comment by danghvu** [comment]
@pavanbalaji @hajimefu I need to clarify the semantic/idea of this refactoring.

According to the original PR, in the end we are going to create these new functions `MPID_Test*/Wait*` which takes the same arguments as the associated `MPI_Test*/Wait*` series except `MPI_Request` is replaced by `MPIR_Request`.

What is the separation of purposes between the "device" layer and the MPI layer ? For example, should the device set the user flag ? should the device set the status / list of statuses ? Because in many cases I found that the `MPIR_*_impl` series i.e. default implementation of `MPID_*` does not do anything with these arguments.

**Comment by halimamer** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by danghvu** [comment]
test:jenkins/ch4/most

**Comment by halimamer** [comment]
test:jenkins/ch3/all

**Comment by halimamer** [comment]
test:jenkins/ch3/tcp

**Comment by halimamer** [comment]
test:jenkins/ch3/most

**Comment by halimamer** [comment]
test:jenkins/ch4/most

**Comment by halimamer** [comment]
@danghvu @pavanbalaji , isn't this performance optimization hiding the real problem? The `master` branch does not process requests in batches; it is still allocating the long array on the heap and traverses it in full. If we change `master` to do batched processing without moving the code around, the performance gap might show up again. Have you tested this @danghvu ?

**Comment by danghvu** [comment]
> Have you tested this @danghvu ?

No I have not tried to batch the original code, I can try later.

**Comment by danghvu** [comment]
Please move the discussion to  #2957
